마켓 등록 기능 구현

ChaYujin opened the pull request:
마켓의 DB접근을 처리할 JPA레파지토리를 등록하였다.
마켓 등록을 처리하는 컨트롤러를 생성하였고, 파라메터의 유효성을 검사하고 유효하지 않은 경우에는 404 bad request를 응답하도록 구현하였다.
마켓 등록 기능의 정상 동작과 잘못된 요청에 대한 응답값들에 대한 테스트코드를 작성하였다.

This closes #7

ChaYujin committed:
#7 - 마켓 도메인 엔티티 및 JPA레파지토리 추가

* 마켓 엔티티 추가
   - DB테이블과 컬럼 매핑되도록 설정
   - Getter와 ToString이 자동생성되도록 설정
   - Date정보가 자동으로 생성되어 DB에 들어갈 수 있도록 @EntityListeners 추가. 이와 관련하여 JPAConfig클래스에 @EnableJpaAuditing도 추가

* JPA레파지토리 추가
   - 마켓 도메인의 DB 처리시 사용할 JPA레파지토리 추가

ChaYujin committed:
#7 - 마켓 등록 기능 구현

* Controller
   - 필수로 필요한 필드값은 생성자를 통해 주입되도록 처리
   - 마켓생성을 담당하는 RequestMapping핸들러 추가
   - MarketDataBindingException 에러 발생시 처리를 담당할 ExceptionHandler추가
   - 파라메터의 유효성 검사를 처리할 Validator지정

* MarketValidator
   - 사용자입력값으로 들어오는 값(name과 url)에 대한 유효성 검사

* MarketControllerTest
   - 정상 케이스, 실패 케이스에 대한 테스트 코드 작성

ChaYujin committed:
#7 - 코드 스타일 맞춤

ModuUiMarketApplication과 ModuUiMarketApplicationTests 클래스의 코드 스타일을 맞춤 (들여쓰기)

ChaYujin committed:
#7 - 유효하지 않은 파라메터 수신에 대한 응답코드 변경

유효하지 않은 파라메터 수신에 대한 응답코드가 500 internal server error 보다는 404 bad request 가 더 적합하다고 생각되어 응답 코드를 변경함

chanwoothecali commented on src/main/java/com/flab/modu/domain/market/controller/MarketController.java:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  @PostMapping("/market")
+  public Market createMarket(@Valid @RequestBody Market market, BindingResult bindingResult)
```
컨트롤러에서 바로 엔티티를 받기보다는 DTO로 받는 게 좋다고 생각합니다.

ChaYujin replied:
> 컨트롤러에서 바로 엔티티를 받기보다는 DTO로 받는 게 좋다고 생각합니다.

현재 MarketController코드에는 DTO로 받도록 처리되어있습니다!

chanwoothecali commented on src/main/java/com/flab/modu/domain/market/domain/Market.java:
```diff
@@ -0,0 +1,91 @@
+
+  @Setter
+  @Column(nullable = false, length = 50)
+  @NotNull
```
Market 클래스가 Entity와 DTO의 역할을 동시에 수행하고 있어서 분리가 필요합니다. Validation Annotation은 DTO로 분리되어야 할 것 같습니다.

ChaYujin replied:
> Market 클래스가 Entity와 DTO의 역할을 동시에 수행하고 있어서 분리가 필요합니다. Validation Annotation은 DTO로 분리되어야 할 것 같습니다.

이 부분도 현재 MarketControlller에 적용이 되어있습니다. MarketDto.CreateReqeust클래스가 마켓 생성에 대한 Dto를 수행하고 있고 validation도 체크하는 역할을 하고 있습니다. 그리고 응답을 처리하는 DTO도 생성하여 MarketDto.CreateResponse 가 마켓 생성에 대한 응답값을 처리하고 있습니다.

f-lab-michael commented on src/main/java/com/flab/modu/domain/market/controller/MarketController.java:
```diff
@@ -0,0 +1,59 @@
+    this.marketRepository = marketRepository;
+  }
+
+  @ExceptionHandler
+  public ResponseEntity errorHandler(MarketDataBindingException e) {
+
+    return ResponseEntity.badRequest().body(
+        "{\"message\":\"" + e.getBindingResult().getAllErrors().get(0).getDefaultMessage() + "\"}"
+    );
+  }
```
에러 처리 관련된 코드들은 ControllerAdvice 를 사용해서 별도 클래스에서 관리하는게 어떨까요?

ChaYujin replied:
공통Advice로 빼서 처리할 생각을 못했습니다. ControllerAdvice를 적용하는 방향으로 수정하겠습니다!

f-lab-michael commented on src/main/java/com/flab/modu/domain/market/domain/Market.java:
```diff
@@ -0,0 +1,91 @@
+    return new Market(sellerId, name, url);
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof Market)) {
+      return false;
+    }
+    Market market = (Market) o;
+    return id.equals(market.id);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(id);
+  }
```
아이디 기준으로 equals 와 hashcode 를 제공하는 이유에 대해서 설명해주세요

ChaYujin replied:
아아디 기준으로 equals와 hashcode를 만들 당시에.... 만들었던 이유는 2가지가 있었습니다.

1. 클론코딩 했던 프로젝트에서 엔티티에 만들어서 따라서 만들었습니다.
2. equals와 hashCode를 오버라이딩할 때 의미있는 값으로 오버라이딩해야한다고 알고 있었습니다. 그래서 id기준으로 내용을 작성하였습니다.

이러한 이유로 equals와 hascode를 오버라이딩하였는데, 클론코딩했던 프로젝트 내용을 다시 확인해보니 해당 프로젝트에서는 엔티티를 컬렉션에 담아 처리하기때문에 작성한 것이었습니다.
Market엔티티는 당장에 컬렉션에 담아 처리할 상황이 없기때문에 없애야할 것 같습니다.

f-lab-michael replied:
일단 id기준으로 하는걸 없애실 필요는 없습니다. 
하지만 이렇게 하는 이유를 잘 이해하고 사용하실 필요는 있습니다.
JPA에서는 엔티티의 아이디를 기준으로 같은 엔티티인지 여부를 비교하기 때문에, 
Java application 에서도 id기준으로 equals 와 hashcode 를 정의하면 개발하는 과정에서 편하게 사용할수있습니다.
예를들어 읽어온 엔티티와 입력받은 엔티티가 동일한 엔티티인지를 비교하기 위해 id기준으로 equals를 구현했다면
`markget.equals(inputMarket)` 으로 할수있지만 그렇지 않다면 market.getId().equals(inputMarket.getId()) 형태로 해야 합니다. inputMarket 에 세팅된 모든 필드가 읽어온 market 과 일치할지는 보장하기 어렵기 때문입니다.

f-lab-michael commented on src/main/java/com/flab/modu/domain/market/domain/MarketValidator.java:
```diff
@@ -0,0 +1,41 @@
+      errors.rejectValue("name", "market.name.invalid", "too long name");
+    }
+
+    Pattern urlPattern = Pattern.compile("^[a-zA-Z]*$");
```
market 의 url 패턴이 영문자만 쓸수있는건가요?

ChaYujin replied:
넵! 깃허브의 개인 레파지토리처럼 처럼 판매자의 마켓 고유주소를 직접 지정할 수 있는 기능으로 생각했습니다. 그래서 영문자 대소문자만 입력할 수 있도록 처리하였습니다.

f-lab-michael commented on src/main/java/com/flab/modu/domain/market/controller/MarketController.java:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  @PostMapping("/market")
+  public Market createMarket(@Valid @RequestBody Market market, BindingResult bindingResult)
```
엔티티를 controller 에 노출하는건 좋지 않은 방법입니다.
dto 와 관련된 리서치를 해보시고 dto를 적용해주세요

f-lab-michael commented on src/main/java/com/flab/modu/domain/market/domain/MarketValidator.java:
```diff
@@ -0,0 +1,41 @@
+import org.springframework.validation.Validator;
+
+@Component
+public class MarketValidator implements Validator {
```
Validator 를 적용하는 방법도 좋지만 위에 얘기한 DTO를 도입하면서 필드별로 spring-boot-starter-validation에서 제공하는 어노테이션을 적용하는것도 더 쉽고 가독성이 좋습니다. 
`@NotBlank` `@Size` 등등

f-lab-michael commented on src/main/java/com/flab/modu/domain/market/controller/MarketController.java:
```diff
@@ -0,0 +1,59 @@
+      throws MarketDataBindingException {
+    //TODO seller_id는 현재 파라메터로 받게 되어있으나, 사용자 인증 부분이 개발이 완료되면, 세션정보에서 가져오도록 수정이 필요하다.
+
+    if (bindingResult.hasErrors()) {
+      throw new MarketDataBindingException(bindingResult);
+    }
+
+    market.setStatus(MarketStatus.ACTIVE);
+    return marketRepository.save(market);
```
비지니스 로직들은 service layer 를 별도로 구성하여 그 안에서 처리하는게 어떨까요?

f-lab-michael commented on src/main/java/com/flab/modu/domain/market/exception/MarketDataBindingException.java:
```diff
@@ -0,0 +1,14 @@
+import lombok.Getter;
+import org.springframework.validation.BindingResult;
+
+public class MarketDataBindingException extends Exception {
```
엔티티별로 익셉션을 각각 만드실 계획인가요?

ChaYujin replied:
> 엔티티별로 익셉션을 각각 만드실 계획인가요?

엔티티에 해당되는 패키지에 각각 발생할 수 있는 익셉션을 만들어 관리하기로 하였습니다. 
예를 들어, 마켓에 한해서는 마켓주소가 중복인 것에 대한 익셉션을 만들어서 처리할 예정입니다.

MarketDataBindingException은 찬우님과 컨벤션을 맞추기 전에 만든 익셉션이여서 변경이 필요합니다. 
DataBinding에 대한 익셉션은 모든 Handler에 발생할 수 있는 공통 익셉션으로 GlobalExceptionHandler안에 공통으로 처리할 예정이여서 MarketDataBindingException은 제거될 것 같습니다.

f-lab-michael commented on src/main/java/com/flab/modu/domain/market/controller/MarketController.java:
```diff
@@ -0,0 +1,59 @@
+    binder.addValidators(marketValidator);
+  }
+
+  @PostMapping("/market")
```
resource 는 복수형으로 쓰는것이 더 좋습니다. 
`복수형리소스`/`리소스아이디` 로 표현이 되기 때문에 
GET `/margets` 하면 마켓 전체 리스트
GET `/margets/1` 1번 마켓 정보와 같이 표현이 되기 때문입니다.

ChaYujin replied:
하나의 마켓을 생성하기때문에 단수형으로 사용했었는데, POST인 경우에도 복수로 통일해서 사용하는 것이 좋군요. 복수형으로 변경해두겠습니다.

f-lab-michael commented on src/test/java/com/flab/modu/domain/market/controller/MarketControllerTest.java:
```diff
@@ -0,0 +1,159 @@
+
+  @Autowired
+  private ObjectMapper objectMapper;
+
```
```java
@Mock
private MarketRepository repository;
//given
//MockitoBDD.given
given(repository.save(any()).willReturn(new Market());


//when- mockmvc 호출

//then
Market expected = new Market(blarblar);
then(repository).should().save(expected);
```

ChaYujin committed:
#7 - Exception을 처리할 클래스 추가

@ControllerAdvice 애노테이션을 사용하여 Exception을 처리할 ControllerExceptionHandler 클래스를 추가하고, 기존에 MarketController에서만 적용되어있던 ExceptionHandler는 제거함

ChaYujin committed:
#7 - 마켓 생성 URL을 복수형으로 변경

마켓 생성시 URL로 표현하는 리소스는 복수형으로 통일하는 것이 좋다. 그러므로 마켓 생성 URL을  POST /market -> POST /markets 로 변경한다.

ChaYujin committed:
#7 - MarketService 추가

마켓의 비즈니스 로직을 처리할 서비스 클래스 추가. Controller에서 마켓 생성 요청 수신 후에 비즈니스 로직은 서비스에서 처리하도록 수정함.

ChaYujin committed:
#7 - 코드스타일 컨벤션에 맞게 탭사이즈 변경 (2칸->4칸)

구글 코드 스타일을 바탕으로 탭 사이즈만 4칸으로 바꿔 사용하기로 결정하여 4칸으로 변경하여 적용함

ChaYujin committed:
#7 - MarketController에 요청파라메터, 응답값에 DTO로 처리하도록 변경

- 컨트롤러에 엔티티가 직접 노출이 되면 원치않는 값이 들어가거나 변경되거나 노출될 위험성이 있기때문에 엔티티대신에 요청과 응답을 처리할 DTO를 생성함.
- Market과 연관되어있는 DTO 클래스가 많이 생성될 것 같아 MarketDto안에 static nested class로 선언함.
- MarketDto.CreateRequest에 파라메터 유효성 검사를 하기 위해 어노테이션을 사용함. 이를 적용함에 따라 불필요해진 MarketValidator클래스를 제거함.
- Market엔티티에 불필요하다고 생각되는 Setter를 제거하였고, Builder패턴에 MarketStatus값을 추가함.

ChaYujin committed:
#7 - Exception을 처리하는 ControllerAdvice클래스명과 패키지 변경

Exception을 처리할 ControllerAdvice어노테이션이 붙은 클래스의 이름을 ControllerExceptionHander->GlobalExceptionHandler로 변경함.
패키지도 advice->exception으로 변경함.

f-lab-michael commented on src/main/java/com/flab/modu/domain/market/controller/MarketController.java:
```diff
@@ -0,0 +1,33 @@
+            throw new MarketDataBindingException(bindingResult);
+        }
+
+        Market market = marketService.createMarket(createMarketRequest.toEntity());
```
컨벤션으로 정해야 하는 문제인데 dto -> entity 의 변환을 controller 에서 할지 service 할지 통일시키면 좋습니다.
보통 service 영역에서 많이 합니다.

f-lab-michael commented on src/test/java/com/flab/modu/domain/market/controller/MarketControllerTest.java:
```diff
@@ -0,0 +1,159 @@
+    String url = "";
+    for (int i = 0; i < 10; i++) {
+      url += "ThisIsUrls";
+    }
```
참고로 RandomStringUtils 라는 유틸클래스를 apache common langs 에서 제공합니다.

ChaYujin committed:
#7 - 테스트코드의 검증 문자열을 RandomStringUtils를 사용하도록 수정

최대 길이의 입력값과 최대 길이를 초과한 입력값에 대한 테스트의 문자열을 하드코딩이 아닌 RandomStringUtils를 이용하여 만들도록 수정함.
RandomStringUtils사용을 위해 의존성도 추가함.

ChaYujin committed:
#7 - Dto -> Entity 변환을 서비스에서 하도록 수정

Controller단에서 Dto -> Entity 변환하지 않고 Service에서 변환하도록 코드 수정함

ChaYujin committed:
#7 - MethodArgumentNotValidException을 처리하는 Exception Handler 추가 및  Dto Validation message추가

- Dto 파라메터 DataBinding시 유효하지않은 파라메터로 오류가 발생하는 경우 MethodArgumentNotValidException이 발생함. 이 때문에 MethodArgumentNotValidException을 공통으로 처리할 ExceptionHandler를 GlobalExceptionHandler 클래스에 추가함.
- 불필요해진 MarketDataBidingException클래스를 제거하고, MarketController 코드를 수정함.
- 유효하지않은 파라메터로 MarketDto DataBinding시 오류 발생했을 경우 메세지를 표시하기위해 valid어노테이션에 message속성 추가

ChaYujin committed:
#7 - 패키지 구조 통일

회원가입 기능의 도메인 패키지구조와 마켓생성 기능의 패키지 구조가 달라 구조를 통일하였다.
마켓에 대한 패키지 구조를 com.flab.modu.domain.market->com.flab.modu.market으로 변경함.

f-lab-michael commented on src/main/java/com/flab/modu/market/domain/Market.java:
```diff
@@ -0,0 +1,84 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(nullable = false, length = 50)
+    @NotNull
```
@Column(nullable = false)와 @NotNull 은 같은효과를 내서 하나만 적용해도 되지 않을까요?

ChaYujin replied:
> @column(nullable = false)와 @NotNull 은 같은효과를 내서 하나만 적용해도 되지 않을까요?

@Column의 nullable과 @NotNull이 다른 용도로 사용된다고 생각합니다. 
@Column의 nullable의 DB에 null을 허용하지 않기 위해 사용하고, @NotNull은 validation 으로 유효성을 검사할 때 사용하는 것으로 알고 있습니다.
전에 위의 코드가 함께 들어가 있는 것은 전에 Entity를 컨트롤러에 노출시켰을 때 짰던 코드입니다. Entity와 DTO를 분리하면서 미처 제거하지 못했네요. 저 Entity클래스에서 @NotNull을 없애고 @Column만 쓰면 될 것 같습니다.

내용 추가합니다!
다시 찾아보니 @NotNull에서 제가 알지 못한 기능이 있었네요. @NotNull을 사용해도 DDL에 not null이 들어가고, Hibernate validation에서 validation체크하는 기능이 있어서 쿼리 실행 전에 validation을 검사하여 사전에 null로 요청하는 것을 방지할 수 있네요. 
그러면 @Column(nullable) 보다 @NotNull을 사용하는 것이 더 나을까요?

f-lab-michael commented on src/test/java/com/flab/modu/market/controller/MarketControllerTest.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    @Test
+    public void failCreatingMarketByWrongParams() throws Exception {
```
하나의 테스트 메소드에 한번의 테스트가 있는것이 좋습니다.
잘못된 여러가지 파라미터를 테스트하고 싶은 니즈면 `@ParameterizedTest` 를 사용해보세요

f-lab-michael commented on src/test/java/com/flab/modu/market/controller/MarketControllerTest.java:
```diff
@@ -0,0 +1,174 @@
+import org.springframework.test.web.servlet.MockMvc;
+
+@ExtendWith(SpringExtension.class)
+@SpringBootTest
```
의도적으로 통합테스트를 하신걸까요?

ChaYujin replied:
> 의도적으로 통합테스트를 하신걸까요?

아니요! 현재 컨트롤러, 서비스, 레파지토리 레이어별로 단위테스트를 하도록 테스트코드를 전면 수정하고 있습니다. 통합테스트가 아닌 단위테스트를 수행하도록 수정하여 올리겠습니다.

ChaYujin committed:
#7 - 마켓 생성시 마켓주소 중복 등록 방지 처리 및 Service layer 테스트코드 추가

- 마켓 고유 속성중에 하나인 마켓주소는 중복등록할 수 없으므로 주소중복 확인 기능을 추가함
- 마켓 주소 중복의 경우 DuplicatedUrlException을 발생시킴
- Service layer의 테스트코드를 추가하고, 마켓생성 성공 케이스와 마켓 주소 중복으로 인한 생성 실패 케이스를 작성함

ChaYujin committed:
#7 - 마켓 JPA 테스트 추가

repository레이어의 테스트코드를 추가함

ChaYujin committed:
#7 - 서비스 단위테스트에 테스트 이름 추가

서비스 단위테스트 테스트 결과에 테스트 이름이 표시되도록 DisplayName추가함

ChaYujin committed:
#7 - Market엔티티에 속성업데이트 메소드 추가

마켓 JPA 단위테스트 코드에서 업데이트 테스트에 필요한 속성업데이트 메소드 추가함.
updateMarket()메소드로 마켓명, 마켓주소, 마켓상태를 업데이트할 수 있음.

ChaYujin committed:
#7 - 마켓 컨트롤러 테스트를 단위테스트로 수정

- MarketControllerTest가 통합테스트가 아닌 단위테스트로 동작하도록 @SpringBootTest를 제거하고, @WebMvcTest애노테이션을 달아주었음
-  컨트롤러 레이어의 코드의 단위테스트를 위해 컨트롤러에서 호출하는 MarketService를 mockito로 mocking하여 사용
- API 문서작업을 위해 정상케이스 테스트 성공시 도큐먼트 를 작성하는 코드 추가
- 컨트롤러에서 엔티티 -> 응답DTO로 변환하는 작업을 서비스에서 수행하도록 수정

ChaYujin committed:
#7 - 엔티티에 @NotNull 어노테이션 제거

@NotNull은 컨트롤러에 노출되어있을 당시 validation체크를 위해 추가해놨던 어노테이션으로 현재는 MarketDto에서 그 역할을 수행하므로 불필요한 어노테이션으로 판단되어 제거함
